### PR TITLE
#165413099 intergrate jwt

### PR DESCRIPTION
**Pull request by ericnyirimana**
Description
=======
Every system application needs authentication to permit users to access their information. With this PR I have integrated JWT( JSON WebToken ) for generating a token made of encrypted user information for session checking and privileges.

Type of change
=======
N/A

How has it been tested
=======
- It is tested when you signup via ```api/v1/users/signup```
- It is tested when you login via ```api/v1/users/login```

Checklist:
=======
N/A

Pivotal tracker story ID
=======
[165413099](https://www.pivotaltracker.com/n/projects/2326984/stories/165413099)

### CODE REVIEW

File: helpers/index.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    return hashedPassword;
+  }
+  static comparePassword(passwordHash, password) {
+    const comparedPassword = bcrypt.compareSync(password, passwordHash);
+    return comparedPassword;
+  }
+
+}
+
+export default Helper;
```
Comment (Kabalisa) [inline helpers/index.js]: Add a new line

File: routes/user.route.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import express from 'express';
+
+import addUser from '../controllers/user.controller';
+
+const router = express.Router();
+
+router.post('/signup', addUser);
+
+export default router;
```
Comment (Kabalisa) [inline routes/user.route.js]: also add a new line here

File: controllers/user.controller.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      };
+      const issueToken = Helper.generateToken(userData);
+      return res.status(201).send({ status: 201, message: 'User added successfully', token: issueToken, data: userData });
+    }
+    if(err){
+      return res.status(409).send({ status: 409, message: err.message });
+    }
+  })(req, res)
+};
+export default addUser;
```
Comment (Kabalisa) [inline controllers/user.controller.js]: add a new line here

**Review by Kabalisa — APPROVED** [review]
@ericnyirimana Good job. please update your PR as it currently has conflicts with develop.

**Review by Niyitangasam — APPROVED** [review]

File: tests/user.test.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+};
+
+describe('Users Authentication', () => {
+    it('User Signup', (done) => {
+      chai.request(server)
+        .post('/api/users/signup')
+        .send(user)
+        .end((err, res) => {
+            expect(res.body).to.have.status(201);
+            expect(res.body.data.username).to.be.equal(user.username);
```
Comment (macphilips) [inline tests/user.test.js]:
```suggestion
            expect(res.body.data.username).to.be.equal(user.username);
            expect(res.body.data.password).to.be.equal(undefined);
```

**Review by macphilips — CHANGES_REQUESTED** [review]

File: package.json [hunk 5]
```diff
@@ -4,33 +4,35 @@
   "description": "A Social platform for the creative at heart",
   "main": "index.js",
   "scripts": {
-    "test": "nyc mocha --require @babel/register \"tests/**/*.js\"",
+    "test": "NODE_ENV=test npm run migrate && NODE_ENV=test  nyc mocha --require @babel/register --require @babel/polyfill \"tests/**/*.js\"",
```
Comment (macphilips) [inline package.json]:
```suggestion
     "pretest": "npm run lint:fix && NODE_ENV=test npm run migrate && NODE_ENV=test",
    "test": "NODE_ENV=test  nyc mocha --require @babel/register --require @babel/polyfill \"tests/**/*.js\"",
```

**Review by macphilips — CHANGES_REQUESTED** [review]

**Pull request merged**